components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java

```diff
@@ -270,4 +283,114 @@ private static boolean isUserConsentRequiredForClaims(String grantType) {
+            throws IdentityOAuth2Exception {
+
+        if (SubjectType.PAIRWISE.equals(getSubjectType(clientId))) {
+            String sectorIdentifierUri = getSectorIdentifierUri(clientId);
+            return getPairwiseSubjectIdentifier(sectorIdentifierUri, authenticatedSubjectIdentifier, callBackURI);
+        }
+        return authenticatedSubjectIdentifier;
+    }
+
+    /**
+     * Get the sector identifier URI from the service provider properties.
+     *
+     * @param clientId client id
+     * @return sector identifier URI
+     * @throws IdentityOAuth2Exception when an error occurred while getting the service provider properties
+     */
+    private static String getSectorIdentifierUri(String clientId) throws IdentityOAuth2Exception {
+
+        ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+        ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+        for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+            if (SECTOR_IDENTIFIER_URI.equals(serviceProviderProperty.getName())) {
+                return serviceProviderProperty.getValue();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Calculate pairwise subject identifier.
+     * <a href="https://openid.net/specs/openid-connect-core-1_0.html#PairwiseAlg">...</a>
+     *
+     * @param sectorIdentifierUri sector identifier URI
+     * @param userId              user id
+     * @param callBackURI         callback URI
+     * @return pairwise subject identifier
+     * @throws IdentityOAuth2Exception if required values are not present
+     */
+    private static String getPairwiseSubjectIdentifier(String sectorIdentifierUri, String userId, String callBackURI)
+            throws IdentityOAuth2Exception {
+
+        URI uri = StringUtils.isNotBlank(sectorIdentifierUri) ? URI.create(sectorIdentifierUri) :
+                StringUtils.isNotBlank(callBackURI) && isValidCallBackURI(callBackURI) ? URI.create(callBackURI) : null;
+        String hostname;
+        if (uri != null) {
+            hostname = uri.getHost();
+        } else {
+            throw new IdentityOAuth2Exception("Invalid sector identifier URI or callback URI.");
+        }
+
+        if (StringUtils.isBlank(userId)) {
+            throw new IdentityOAuth2Exception("Invalid user id.");
+        }
+
+        String seed = hostname.concat(userId);
+        return UUID.nameUUIDFromBytes(seed.getBytes(StandardCharsets.UTF_8)).toString();
+    }
+
+    private static boolean isValidCallBackURI(String callBackURI) {
+
+        return !callBackURI.startsWith(OAuthConstants.CALLBACK_URL_REGEXP_PREFIX);
+    }
+
+    public static String getCallbackUrl(String clientId) throws IdentityOAuth2Exception, InvalidOAuthClientException {
+
+        OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId);
```

Comment:
This method is deprecated now. lets use the method with tenant domain as well. update in other places.